pkg/cosign/tsa.go

```diff
@@ -0,0 +1,161 @@
+	targets, err := tufClient.GetTargetsByMeta(usage, names)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching targets by metadata with usage %v: %w", usage, err)
+	}
+	for _, target := range targets {
+		certs, _, _, err := splitPEMCertificateChain(target.Target)
+		if err != nil {
+			return nil, fmt.Errorf("error splitting PEM certificate chain for target: %w", err)
+		}
+		allCerts = append(allCerts, certs...)
+	}
+	return allCerts, nil
+}
+
+// GetTSACerts retrieves trusted TSA certificates from the embedded or cached
+// TUF root. If expired, makes a network call to retrieve the updated targets.
+// By default, the certificates come from TUF, but you can override this for test
+// purposes by using an env variable `SIGSTORE_TSA_CERTIFICATE_FILE` or a file path
+// specified in `TSACertChainPath`. If using an alternate, the file should be in PEM format.
+func GetTSACerts(ctx context.Context, certChainPath string, fn GetTargetStub) (*TSACertificates, error) {
+	altTSACert := env.Getenv(env.VariableSigstoreTSACertificateFile)
+
+	var raw []byte
+	var err error
+
+	switch {
+	case altTSACert != "":
+		raw, err = os.ReadFile(altTSACert)
+	case certChainPath != "":
+		raw, err = os.ReadFile(certChainPath)
+	default:
+		certNames := []string{tsaLeafCertStr, tsaRootCertStr}
+		for i := 0; ; i++ {
+			intermediateCertStr := fmt.Sprintf(tsaIntermediateCertStrPattern, i)
+			_, err := fn(ctx, tuf.TSA, []string{intermediateCertStr})
+			if err != nil {
+				break
+			}
+			certNames = append(certNames, intermediateCertStr)
+		}
+		allCerts, err := fn(ctx, tuf.TSA, certNames)
+		if err != nil {
+			return nil, fmt.Errorf("error fetching TSA certificates: %w", err)
+		}
+
+		var concatenatedPEM []byte
+		for _, cert := range allCerts {
+			pemData, err := cryptoutils.MarshalCertificatesToPEM([]*x509.Certificate{cert})
+			if err != nil {
+				return nil, fmt.Errorf("error marshalling certificate to PEM: %w", err)
+			}
+			concatenatedPEM = append(concatenatedPEM, pemData...)
+		}
+
+		leaves, intermediates, roots, err := splitPEMCertificateChain(concatenatedPEM)
+		if err != nil {
+			return nil, fmt.Errorf("error splitting TSA certificates: %w", err)
+		}
+
+		if len(leaves) != 1 {
+			return nil, fmt.Errorf("expected exactly one leaf certificate, found %d", len(leaves))
+		}
+
+		return &TSACertificates{
+			LeafCert:          leaves[0],
+			IntermediateCerts: intermediates,
+			RootCert:          roots,
+		}, nil
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error reading TSA certificate file: %w", err)
+	}
+
+	leaves, intermediates, roots, err := splitPEMCertificateChain(raw)
+	if err != nil {
+		return nil, fmt.Errorf("error splitting TSA certificates: %w", err)
+	}
+
+	if len(leaves) > 1 {
+		return nil, fmt.Errorf("TSA certificate chain must contain at most one TSA certificate")
+	}
+
+	return &TSACertificates{
+		LeafCert:          leaves[0],
+		IntermediateCerts: intermediates,
+		RootCert:          roots,
+	}, nil
+}
```

Comment:
@ianhundere  lines of code 103-116 and lines 123-136 look very similar, maybe delete 103-116 and assign raw = concatenatedPEM?

Reply:
doh, thanks for that. 🙇